clustermesh: cli: use and support dict for clusters helm values

MrFreezeex opened the pull request:
Support `clustermesh.config.clusters` to be a dict and process internally the clusters list as a dict. For now we always output to a list as ReuseValues is true when invoking helm (see the code comment for more details) but will be refactored further to allow outputting a dict conditionally in a future commit.

Related to #40796
Support for helm is added in: #40857

```release-note
clustermesh: cli: add support for dict for helm values `clustermesh.config.clusters`
```

MrFreezeex committed:
clustermesh: cli: use and support dict for clusters helm values

Support `clustermesh.config.clusters` to be a dict and process
internally the clusters list as a dict. For now we always output to a
list as ReuseValues is true when invoking helm (see the code comment for
more details) but will be refactored further to allow outputting a dict
conditionally in a future commit.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1602,36 +1643,61 @@ func (k *K8sClusterMesh) displayDisconnectedCompleteMessage(localClient *k8s.Cli
 }
 
-func getOldClusters(values map[string]any) ([]map[string]any, error) {
+func filterClusters(clusters map[string]any) (map[string]any, error) {
```
nit: getEnabledClusters or getActiveClusters maybe?

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1222,8 +1223,11 @@ func (k *K8sClusterMesh) processSingleRemoteClient(ctx context.Context, remoteCl
+	if state.localNewClusters == nil {
+		state.localNewClusters = make(map[string]any)
+	}
+	state.localNewClusters[newClusterName] = newCluster
```
Should we warn if cluster with this name already existed?

MrFreezeex replied:
I made it hard fail since we have some similar error here: https://github.com/cilium/cilium/blob/021146c542683740a58d5d39ad82fa704ec51af9/cilium-cli/clustermesh/clustermesh.go#L540 It probably should fail here too IMO

marseel commented on cilium-cli/clustermesh/clustermesh_test.go:
```diff
@@ -711,5 +650,147 @@ func TestGetCASecret(t *testing.T) {
+}
+
+func TestGetClustersFromValues(t *testing.T) {
+	uu := map[string]struct {
```
nit: I would make it a list with field "name", similar to `TestGetCASecret`. 
It will be a bit easier to ready with one less indent.

MrFreezeex committed:
clustermesh: cli: fail if multiple remote clusters have the same name

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

marseel commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1222,8 +1223,14 @@ func (k *K8sClusterMesh) processSingleRemoteClient(ctx context.Context, remoteCl
-	state.localNewClusters = append(state.localNewClusters, newCluster)
+	newClusterName, newCluster := getCluster(aiRemote, !match)
+	if state.localNewClusters == nil {
+		state.localNewClusters = make(map[string]any)
```
nit: I would probably initialize `localNewClusters` and `remoteOldClustersAll` (and others) before within `processLocalClient` when state is created. 
You could get rid of these lines, lines below with nil checks for `state.remoteOldClustersAll`, `remoteHelmValuesMesh`and `remoteNewClusters` like:
```
	remoteNewClusters := maps.Clone(state.localNewClusters)
	if remoteNewClusters == nil {
		remoteNewClusters = make(map[string]any)
	}
```
Imho it would be much simpler to read.

MrFreezeex replied:
There was a couple more that we could initialize like that and actually I think it might be better to just initialize them all even if they get overridden after so that we don't have to reason about it and potentially loose time on some code change later on to realize that some specific map wasn't initialized.

This is definitely subjective though because technically we are doing something unnecessary here that could maybe be confusing in a way. So like I can definitely limit that to only the ones with those nil check if you think that's more reasonable!

marseel replied:
All sounds good, thanks!

marseel reviewed (approved):
Thanks, looks good!

MrFreezeex committed:
cilium-cli: initialize all maps on ClusterMesh state creation

Initialize all maps in the ClusterState by default so that we can remove
various nil check to lazy initialize those. Some of those maps are
getting override later too but we are still initializing here to not
have to reason about in potential refactoring later on.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

joestringer commented on cilium-cli/clustermesh/clustermesh.go:
```diff
@@ -1301,7 +1304,40 @@ func (k *K8sClusterMesh) connectLocalWithHelm(ctx context.Context, localClient *
+	// FIXME: force output to list for now. This can be removed conditionally
+	//        with a version check once we would be able to set ReuseValues to
+	//        false
+	values = convertToListClusterMeshValues(values)
```
Are you planning to address this limitation?

MrFreezeex replied:
Yes that's one of my next goal that I wanted to address separately, I would probably send a PR by next week or so for that

MrFreezeex replied:
If I might add technically we could be able to leave the code in this state I just added a fixme here because I will followup shortly on that and it's a bit better to have it but not even required or anything

joestringer replied:
Developers are notorious for leaving landmines in the codebase for future developers to find and not following up. If we are OK with the current implementation, I would suggest just deleting the fixme. It's great if you plan to follow up to improve. What I disagree with is documenting such intentions into the tree: If it's important, we should agree it's important and block this PR until it's resolved. If it's not important, then there's no need to leave a note for our future selves. We can agree the current implementation is fine and then just follow up at your discretion.

joestringer replied:
Above is not a hard rule, but general guidance. If something is OK but could be improved, we have from time to time added context in the code with a comment like `// This does X and could be improved by changing to do Y. See GH-XXXX for more details` and file an issue to follow up.

MrFreezeex replied:
Ah I was about to comment but got sidetracked, thanks for the approval :D

joestringer reviewed (changes requested):
Could this impact existing users in how they are currently specifying the configuration?

One blocking comment below. Typically we don't add TODOs into the tree.
